python-package/xgboost/distributed.py

```diff
@@ -0,0 +1,369 @@
+        y_parts = y_parts.flatten().tolist()
+
+    assert len(X_parts) == len(
+        y_parts), 'Partitions between X and y are not consistent'
+
+    if weights:
+        w_parts = weights.to_delayed()
+        assert len(X_parts) == len(
+            w_parts), 'Partitions between X and weight are not consistent.'
+        parts = list(map(delayed, zip(X_parts, y_parts, w_parts)))
+    else:
+        parts = list(map(delayed, zip(X_parts, y_parts)))
+
+    parts = client.compute(parts)
+    await wait(parts)  # async wait for parts to be computed
+
+    for part in parts:
+        assert part.status == 'finished'
+
+    key_to_partition = dict([(part.key, part) for part in parts])
+    who_has = await client.scheduler.who_has(
+        keys=[part.key for part in parts])
+
+    worker_map = defaultdict(list)
+    for key, workers in who_has.items():
+        worker_map[first(workers)].append(key_to_partition[key])
+    return worker_map
+
+
+def dispatch_fit(func):
+    def dispatcher(X, y, sample_weight=None, *args, **kwargs):
+        client = get_client()
+
+        def dispatcched_fit():
+            worker = distributed_get_worker()
+            list_of_parts = worker_map[worker.address]
+            list_of_parts = [p.result() for p in list_of_parts]
+
+            if sample_weight:
+                data, labels, weights = zip(*list_of_parts)
+            else:
+                data, labels = zip(*list_of_parts)
+
+            data = concat(data)
+            labels = concat(labels)
+            if sample_weight:
+                weights = concat(weights)
+            else:
+                weights = None
+
+            return func(data, labels, weights, *args, **kwargs)
+
+        if type(X) in {DDF, DA}:
+            worker_map = map_local_data(X, y, sample_weight)
+            futures = client.map(dispatcched_fit, X, y, sample_weight,
+                                 *args, **kwargs)
+            return futures
+        else:
+            return func(X, y, sample_weight, *args, **kwargs)
+
+    return dispatcher
+
+
+class DaskDMatrix:
+    __doc__ = '''DMatrix holding on references to Dask DataFrame or ''' + \
+        '''Dask Array.\n\n'''.join(DMatrix.__doc__)
+
+    _feature_names = None  # for previous version's pickle
+    _feature_types = None
+
+    def __init__(self,
+                 X,
+                 y=None,
+                 missing=None,
+                 weight=None,
+                 feature_names=None,
+                 feature_types=None):
+        self._feature_names = feature_names
+        self._feature_types = feature_types
+        self._missing = missing
+
+        assert len(X.shape) == 2, ('Expecting 2 dimensions input, got: ' +
+                                   str(X.shape))
+        self.n_rows = X.shape[0]
+        self.n_cols = X.shape[1]
+
+        self.X = X
+        self.y = y
+        self.weights = weight
+
+        assert type(X) in (DDF, DA)
+        assert type(y) in (DDF, DS, DA)
```

Comment:
I had to look these up in the imports above.  I wonder if it might be clearer to use `da.Array` and `dd.DataFrame`.  These are longer, yes, but they're also terms that will be well understood by anyone who uses Dask.

Reply:
@mrocklin Sure.  Will refactor it properly.